File: services/assets-api/app/items.py
"""Asset (CI) writes — create and version-locked patch. Archive rides the
patch ({archived: true/false}); retirement is the 'retired' STATUS (still
listed), archiving hides the row. No DELETE anywhere (house doctrine)."""
from fastapi import APIRouter, HTTPException, Request
from psycopg import errors as pg_errors
from pydantic import BaseModel
from . import auth, db, helpers

router = APIRouter()


class NewAsset(BaseModel):
    ci_tag: str
    name: str
    atype: str
    client_id: str
    assigned_to: str = ""
    status: str = "inuse"
    serial: str = ""
    vendor: str = ""
    purchased_on: str | None = None
    warranty_until: str | None = None
    cost_cents: int | None = None
    note: str = ""


@router.post("/api/assets", status_code=201)
def create_asset(body: NewAsset, request: Request):
    ci_tag, name = body.ci_tag.strip(), body.name.strip()
    if not ci_tag or not name:
        raise HTTPException(422, "CI tag and name are both required")
    if body.atype not in helpers.ATYPES:
        raise HTTPException(422, f"atype must be one of {', '.join(helpers.ATYPES)}")
    if body.status not in helpers.STATUSES:
        raise HTTPException(422, f"status must be one of {', '.join(helpers.STATUSES)}")
    purchased = helpers.sane_date(body.purchased_on, "purchased_on")
    warranty = helpers.sane_date(body.warranty_until, "warranty_until")
    helpers.cents_ok(body.cost_cents)
    with db.connect() as conn:
        who = auth.require(conn, request)
        auth.need(who, "a_manage_assets")
        with conn.cursor() as cur:
            helpers.client_or_422(cur, body.client_id)
            try:
                cur.execute("""INSERT INTO assets.assets
                                 (ci_tag, name, atype, client_id, assigned_to, status,
                                  serial, vendor, purchased_on, warranty_until, cost_cents, note)
                               VALUES (%s,%s,%s,%s,%s,%s,%s,%s,%s,%s,%s,%s)
                               RETURNING id, version, updated_at""",
                            (ci_tag, name, body.atype, body.client_id, body.assigned_to.strip(),
                             body.status, body.serial.strip(), body.vendor.strip(),
                             purchased, warranty, body.cost_cents, body.note.strip()))
            except pg_errors.UniqueViolation:
                raise HTTPException(409, f"CI tag “{ci_tag}” is already in use")
            aid, version, updated_at = cur.fetchone()
            helpers.event(cur, "asset", aid, who,
                          f"Asset added — {ci_tag} {name} ({body.atype})")
        auth.audit(conn, "assets", "Asset added", f"asset:{aid}",
                   f"{ci_tag} {name} · {body.atype} ({who['label']})")
        return {"id": str(aid), "version": version, "updatedAt": helpers.ms(updated_at)}


class PatchAsset(BaseModel):
    version: int                      # optimistic-lock token from the last read
    ci_tag: str | None = None
    name: str | None = None
    atype: str | None = None
    client_id: str | None = None
    assigned_to: str | None = None
    status: str | None = None
    serial: str | None = None
    vendor: str | None = None
    purchased_on: str | None = None   # "" clears
    warranty_until: str | None = None # "" clears
    cost_cents: int | None = None
    clear_cost: bool = False          # cost_cents=None means "untouched", so clearing is explicit
    note: str | None = None
    archived: bool | None = None


@router.patch("/api/assets/{asset_id}")
def patch_asset(asset_id: str, body: PatchAsset, request: Request):
    with db.connect() as conn:
        who = auth.require(conn, request)
        auth.need(who, "a_manage_assets")
        sets, args, notes = [], [], []
        with conn.cursor() as cur:
            if body.ci_tag is not None:
                tag = body.ci_tag.strip()
                if not tag:
                    raise HTTPException(422, "The CI tag can't be blank")
                sets.append("ci_tag = %s"); args.append(tag); notes.append(f"tag → {tag}")
            if body.name is not None:
                nm = body.name.strip()
                if not nm:
                    raise HTTPException(422, "The name can't be blank")
                sets.append("name = %s"); args.append(nm); notes.append(f"name → {nm}")
            if body.atype is not None:
                if body.atype not in helpers.ATYPES:
                    raise HTTPException(422, f"atype must be one of {', '.join(helpers.ATYPES)}")
                sets.append("atype = %s"); args.append(body.atype); notes.append(f"type → {body.atype}")
            if body.client_id is not None:
                helpers.client_or_422(cur, body.client_id)
                sets.append("client_id = %s"); args.append(body.client_id); notes.append("client moved")
            if body.assigned_to is not None:
                sets.append("assigned_to = %s"); args.append(body.assigned_to.strip())
                notes.append(f"assigned → {body.assigned_to.strip() or '—'}")
            if body.status is not None:
                if body.status not in helpers.STATUSES:
                    raise HTTPException(422, f"status must be one of {', '.join(helpers.STATUSES)}")
                sets.append("status = %s"); args.append(body.status); notes.append(f"status → {body.status}")
            if body.serial is not None:
                sets.append("serial = %s"); args.append(body.serial.strip()); notes.append("serial edited")
            if body.vendor is not None:
                sets.append("vendor = %s"); args.append(body.vendor.strip()); notes.append("vendor edited")
            if body.purchased_on is not None:
                d = helpers.sane_date(body.purchased_on, "purchased_on")
                sets.append("purchased_on = %s"); args.append(d)
                notes.append(f"purchased → {d or '—'}")
            if body.warranty_until is not None:
                d = helpers.sane_date(body.warranty_until, "warranty_until")
                sets.append("warranty_until = %s"); args.append(d)
                notes.append(f"warranty → {d or '—'}")
            if body.cost_cents is not None or body.clear_cost:
                c = None if body.clear_cost else helpers.cents_ok(body.cost_cents)
                sets.append("cost_cents = %s"); args.append(c)
                notes.append("cost cleared" if c is None else f"cost → {c}c")
            if body.note is not None:
                sets.append("note = %s"); args.append(body.note.strip()); notes.append("note edited")
            if body.archived is not None:
                sets.append("archived_at = " + ("now()" if body.archived else "NULL"))
                notes.append("archived" if body.archived else "restored")
            if not sets:
                return {"ok": True, "changed": []}
            try:
                cur.execute(f"""UPDATE assets.assets SET {', '.join(sets)}
                                 WHERE id::text = %s AND version = %s
                               RETURNING ci_tag, version, updated_at""",
                            (*args, asset_id, body.version))
            except pg_errors.UniqueViolation:
                raise HTTPException(409, "That CI tag is already in use")
            row = cur.fetchone()
            if row is None:
                raise HTTPException(409, "Version conflict or missing asset — re-read and retry")
            tag, version, updated_at = row
            helpers.event(cur, "asset", asset_id, who, "Asset updated — " + " · ".join(notes))
        auth.audit(conn, "assets", "Asset updated", f"asset:{asset_id}",
                   f"{tag} · " + " · ".join(notes) + f" ({who['label']})")
        return {"ok": True, "changed": notes, "version": version,
                "updatedAt": helpers.ms(updated_at)}

File: services/assets-api/webui/js/assets/views/items.js
/* ==========================================================================
   views/items.js — the Assets (CI) page and its action functions.
   Endpoints: POST /api/assets · PATCH /api/assets/{id} (version-locked).
   One function per control: optimistic local mutate + diff-guard + fetch +
   oops() rollback. Archive hides a row (archived seg shows them); 'retired'
   is a lifecycle STATUS and stays listed.
   ========================================================================== */

function assetRows(){
  const f=state.af, q=(f.q||'').toLowerCase();
  let rows = f.status==='archived' ? state.assets.filter(a=>a.archived)
           : liveAssets().filter(a=> f.status==='all'||a.status===f.status);
  if(f.client.length) rows=rows.filter(a=>f.client.includes(a.clientId));
  if(f.type.length)   rows=rows.filter(a=>f.type.includes(a.atype));
  if(f.cov!=='all')   rows=rows.filter(a=> f.cov==='none' ? !a.warrantyUntil
                                         : (!!a.warrantyUntil && covClass(a.warrantyUntil)===f.cov));
  if(f.from||f.to)    rows=rows.filter(a=>inDateRange(a.purchasedOn, f.from, f.to));
  if(q) rows=rows.filter(a=>(a.ciTag+' '+a.name+' '+a.assignedTo+' '+clientName(a.clientId)+' '+a.serial+' '+a.vendor).toLowerCase().includes(q));
  return rows;
}

function afClear(){ state.af={status:state.af.status, q:state.af.q, client:[], type:[], cov:'all', from:'', to:''}; render(); }

function viewAssetsPage(){
  const f=state.af;
  const counts={all:liveAssets().length, archived:state.assets.filter(a=>a.archived).length};
  Object.keys(ASSET_STATUS).forEach(k=>counts[k]=liveAssets().filter(a=>a.status===k).length);
  const seg=[['all','All'],['inuse','In use'],['stock','In stock'],['repair','In repair'],['retired','Retired'],['archived','Archived']]
    .map(([k,l])=>`<button class="${f.status===k?'on':''}" onclick="state.af.status='${k}';render()">${l}<span class="pip">${counts[k]||0}</span></button>`).join('');
  const money=canSeeCosts();
  /* filter options: archived clients ride along so their assets stay findable
     (row-37 doctrine: keep them out only when nothing references them) */
  const cOpts=state.clients.filter(c=>!c.sentinel).map(c=>({v:c.id,label:c.name+(c.archived?' (archived)':'')}));
  const tOpts=ATYPES.map(t=>({v:t,label:t}));
  const filtered = f.client.length||f.type.length||f.cov!=='all'||f.from||f.to;
  const pg=paginate('assetsList', assetRows());
  return `
  <div class="toolbar">
    <div class="seg">${seg}</div>
    <div class="search"><span>${icon(IC.search)}</span><input type="search" data-fkey="assets-q" placeholder="Search assets…" value="${esc(f.q)}" oninput="state.af.q=this.value;render()"></div>
    <div class="spacer"></div>
    ${can('a_export_csv')?`<button class="btn" onclick="exportAssetsCSV()">Export CSV</button>`:''}
    ${can('a_manage_assets')?`<button class="btn primary" onclick="assetModal('')">${icon(IC.plus)} Add asset</button>`:''}
  </div>
  <div class="toolbar" style="margin-top:-4px">
    <span style="display:inline-block;min-width:190px;vertical-align:middle">${multiCombo('aft-client', cOpts, f.client, v=>mfToggle('af','client',v), 'All clients')}</span>
    <span style="display:inline-block;min-width:170px;vertical-align:middle">${multiCombo('aft-type', tOpts, f.type, v=>mfToggle('af','type',v), 'All types')}</span>
    <select data-fkey="af-cov" onchange="state.af.cov=this.value;render()" title="Warranty health">
      <option value="all" ${f.cov==='all'?'selected':''}>Warranty: any</option>
      <option value="ok" ${f.cov==='ok'?'selected':''}>Active (beyond ${leadDays()}d)</option>
      <option value="due" ${f.cov==='due'?'selected':''}>Expiring ≤ ${leadDays()}d</option>
      <option value="breach" ${f.cov==='breach'?'selected':''}>Expired</option>
      <option value="none" ${f.cov==='none'?'selected':''}>No warranty on record</option>
    </select>
    <span class="mini muted">purchased</span>
    <input type="date" data-fkey="af-from" value="${esc(f.from)}" onchange="state.af.from=this.value;render()" title="Purchased from">
    <span class="mini muted">–</span>
    <input type="date" data-fkey="af-to" value="${esc(f.to)}" onchange="state.af.to=this.value;render()" title="Purchased to">
    ${filtered?`<button class="btn sm ghost" onclick="afClear()">✕ Clear filters</button>`:''}
    <span class="mini muted" style="margin-left:auto">${pg.total} match${pg.total===1?'':'es'}</span>
  </div>
  <div class="card">
    <table class="tbl">
      <thead><tr><th>CI tag</th><th>Asset</th><th>Type</th><th>Client</th><th>Assigned to</th><th>Status</th><th>Warranty</th>${money?'<th class="num">Cost</th>':''}</tr></thead>
      <tbody>
      ${pg.slice.length?pg.slice.map(a=>`<tr class="clickable" onclick="openAsset('${a.id}')">
        <td><span class="citag">${esc(a.ciTag)}</span></td>
        <td><div class="cell-title">${esc(a.name)}</div><div class="cell-meta">${esc(a.vendor||'—')} · ${esc(a.serial||'no serial')}</div></td>
        <td><span class="type-badge"><span class="tibox">${icon(TYPEIC[a.atype]||IC.assets)}</span>${a.atype}</span></td>
        <td>${esc(clientName(a.clientId))}</td>
        <td>${esc(a.assignedTo||'—')}</td>
        <td>${statusChip(a.status)}${a.archived?' <span class="chip grey slim"><span class="cdot"></span>Archived</span>':''}</td>
        <td>${hdot(a.warrantyUntil)}</td>
        ${money?`<td class="num money">${a.costCents!=null?fmtMoney(a.costCents):'<span class="muted">—</span>'}</td>`:''}</tr>`).join('')
      :`<tr><td colspan="${money?8:7}"><div class="empty">${icon(IC.assets)}<div>No assets match.</div></div></td></tr>`}
      </tbody>
    </table>
    ${pagerBar(pg)}
  </div>`;
}

function exportAssetsCSV(){
  if(!can('a_export_csv')) return;
  const rows=[['CI tag','Name','Type','Client','Assigned to','Status','Serial','Vendor','Purchased','Warranty until','Cost','Note']];
  assetRows().forEach(a=>rows.push([a.ciTag,a.name,a.atype,clientName(a.clientId),a.assignedTo,a.status,
    a.serial,a.vendor,a.purchasedOn||'',a.warrantyUntil||'',
    canSeeCosts()&&a.costCents!=null?(a.costCents/100).toFixed(2):'',a.note]));
  csvDownload('assets.csv', rows);
  toast(rows.length-1+' assets exported.');
}

/* ---- detail modal ---- */
function openAsset(id){
  const a=assetById(id); if(!a) return;
  const cov=contractsCovering(id);
  openModal(`
    <div class="modal-head">
      <span class="tibox" style="width:34px;height:34px;border-radius:8px;background:var(--surface-2);border:1px solid var(--line);display:grid;place-items:center;color:var(--ink-3)">${icon(TYPEIC[a.atype]||IC.assets)}</span>
      <div><h3>${esc(a.name)}</h3><p><span class="citag">${esc(a.ciTag)}</span> · ${a.atype} · ${esc(clientName(a.clientId))}</p></div>
      <button class="x" onclick="closeModal()">×</button>
    </div>
    <div class="modal-body">
      <div class="toolbar" style="margin-bottom:16px">${statusChip(a.status)}${a.archived?'<span class="chip grey slim"><span class="cdot"></span>Archived</span>':''}<span class="hdot ${covClass(a.warrantyUntil)}"><span class="sdot"></span>Warranty ${covText(a.warrantyUntil)}</span></div>
      <div class="detail-grid">
        <div><div class="k">Assigned to</div><div class="v">${esc(a.assignedTo||'—')}</div></div>
        <div><div class="k">Vendor</div><div class="v">${esc(a.vendor||'—')}</div></div>
        <div><div class="k">Serial</div><div class="v mono">${esc(a.serial||'—')}</div></div>
        <div><div class="k">Purchased</div><div class="v">${a.purchasedOn?fmtDate(a.purchasedOn):'—'}</div></div>
        ${canSeeCosts()?`<div><div class="k">Purchase cost</div><div class="v money">${a.costCents!=null?fmtMoney(a.costCents):'—'}</div></div>`:''}
        <div><div class="k">Warranty ends</div><div class="v">${a.warrantyUntil?fmtDate(a.warrantyUntil):'—'}</div></div>
      </div>
      ${a.note?`<div class="line-sep"></div><div class="k" style="margin-bottom:4px">Note</div><div style="font-size:13px;color:var(--ink-2)">${esc(a.note)}</div>`:''}
      ${cov.length?`<div class="line-sep"></div><div class="notice lock">${icon(IC.contracts)}<div>Covered by <b>${esc(cov[0].vendor)}</b> — ${termLabel(cov[0].termMonths,cov[0].recurring)}${canSeeCosts()?' · '+fmtMoney(cov[0].costCents)+'/term':''}, ends ${cov[0].endsOn?fmtDate(cov[0].endsOn):'—'}.${cov.length>1?' (+'+(cov.length-1)+' more)':''}</div></div>`:''}
      <div class="line-sep"></div>
      <div class="k" style="margin-bottom:8px">Changes</div>
      ${eventFeed('asset', id)}
    </div>
    <div class="modal-foot">
      ${can('a_manage_assets')?`<button class="btn ${a.archived?'':'danger'}" onclick="toggleArchiveAsset('${a.id}')">${a.archived?'Restore':'Archive'}</button>`:''}
      <button class="btn" onclick="closeModal()">Close</button>
      ${can('a_manage_assets')?`<button class="btn primary" onclick="assetModal('${a.id}')">Edit asset</button>`:''}
    </div>`, true);
}

function toggleArchiveAsset(id){
  const a=assetById(id); if(!a||!can('a_manage_assets')) return;
  const to=!a.archived;
  closeModal();
  a.archived=to; log(to?'Asset archived':'Asset restored', a.ciTag+' '+a.name, id); render();
  post('PATCH','/api/assets/'+id,{version:a.version, archived:to})
    .then(d=>{ if(d){ if(d.version) a.version=d.version; toast(`${a.ciTag} ${to?'archived':'restored'}.`); } });
}

/* ---- add / edit modal ---- */
function assetModal(id, presetClient){
  if(!can('a_manage_assets')) return;
  const a=id?assetById(id):null;
  const cOpts=pickableClients().map(c=>({v:c.id,label:c.name}));
  /* an edited record whose client was archived must still SHOW its client
     (a blank picker misread as unset — review finding) */
  if(a && !cOpts.some(o=>o.v===a.clientId)) cOpts.unshift({v:a.clientId, label:clientName(a.clientId)+' (archived)'});
  openModal(`
    <div class="modal-head"><div><h3>${a?'Edit asset — '+esc(a.ciTag):'Add asset'}</h3>
      <p>${a?'Version-locked: if someone else edited this asset first, the save is refused — reopen and retry.':'A CI tag is the asset’s handle across the suite (e.g. LT-0451).'}</p></div>
      <button class="x" onclick="closeModal()">×</button></div>
    <div class="modal-body">
      <div class="detail-grid">
        <div class="field"><label>CI tag</label><input type="text" id="af-tag" value="${esc(a?a.ciTag:'')}" placeholder="LT-0001"></div>
        <div class="field"><label>Type</label><select id="af-type">${ATYPES.map(t=>`<option ${a&&a.atype===t?'selected':''}>${t}</option>`).join('')}</select></div>
      </div>
      <div class="field"><label>Name</label><input type="text" id="af-name" value="${esc(a?a.name:'')}" placeholder="Dell Latitude 7440"></div>
      <div class="detail-grid">
        <div class="field"><label>Client</label>${combo('af-client', cOpts, a?a.clientId:(presetClient||''), null, 'Pick a client…')}</div>
        <div class="field"><label>Assigned to</label><input type="text" id="af-user" value="${esc(a?a.assignedTo:'')}" placeholder="Person, rack, room…"></div>
        <div class="field"><label>Status</label><select id="af-status">${Object.entries(ASSET_STATUS).map(([k,v])=>`<option value="${k}" ${a&&a.status===k?'selected':''}>${v[1]}</option>`).join('')}</select></div>
        <div class="field"><label>Serial</label><input type="text" id="af-serial" value="${esc(a?a.serial:'')}"></div>
        <div class="field"><label>Vendor</label><input type="text" id="af-vendor" value="${esc(a?a.vendor:'')}"></div>
        <div class="field"><label>Purchased</label><input type="date" id="af-purch" value="${a&&a.purchasedOn?a.purchasedOn:''}"></div>
        <div class="field"><label>Warranty until</label><input type="date" id="af-warr" value="${a&&a.warrantyUntil?a.warrantyUntil:''}"></div>
        ${canSeeCosts()?`<div class="field"><label>Purchase cost ($)</label><input type="number" id="af-cost" min="0" step="0.01" value="${a&&a.costCents!=null?(a.costCents/100):''}" placeholder="—"></div>`:''}
      </div>
      <div class="field"><label>Note</label><textarea id="af-note" rows="2">${esc(a?a.note:'')}</textarea></div>
    </div>
    <div class="modal-foot"><button class="btn" onclick="closeModal()">Cancel</button>
      <button class="btn primary" onclick="saveAsset('${a?a.id:''}')">${a?'Save asset':'Add asset'}</button></div>`, true);
  document.getElementById(a?'af-name':'af-tag').focus();
}

function saveAsset(id){
  const g=x=>document.getElementById(x);
  const tag=g('af-tag').value.trim(), name=g('af-name').value.trim(), clientId=g('af-client').value;
  if(!tag||!name){ toast('CI tag and name are both required.'); return; }
  if(!clientId){ toast('Pick a client.'); return; }
  const costEl=g('af-cost');
  const costRaw=costEl?costEl.value.trim():'';
  const costCents=costRaw===''?null:Math.round(parseFloat(costRaw)*100);
  if(costCents!=null && (isNaN(costCents)||costCents<0)){ toast('Cost must be a positive amount.'); return; }
  const vals={ ci_tag:tag, name, atype:g('af-type').value, client_id:clientId,
    assigned_to:g('af-user').value.trim(), status:g('af-status').value,
    serial:g('af-serial').value.trim(), vendor:g('af-vendor').value.trim(),
    purchased_on:g('af-purch').value||'', warranty_until:g('af-warr').value||'',
    note:g('af-note').value.trim() };
  closeModal();
  if(!id){
    if(costEl){ if(costCents!=null) vals.cost_cents=costCents; }
    if(!vals.purchased_on) delete vals.purchased_on;
    if(!vals.warranty_until) delete vals.warranty_until;
    log('Asset added', tag+' '+name);
    post('POST','/api/assets',vals).then(d=>{ if(d) toast(`Asset ${tag} added.`); });
    return;
  }
  const a=assetById(id); if(!a) return;
  /* PATCH only what changed (diff-guard, row 21) */
  const body={version:a.version};
  if(vals.ci_tag!==a.ciTag) body.ci_tag=vals.ci_tag;
  if(vals.name!==a.name) body.name=vals.name;
  if(vals.atype!==a.atype) body.atype=vals.atype;
  if(vals.client_id!==a.clientId) body.client_id=vals.client_id;
  if(vals.assigned_to!==a.assignedTo) body.assigned_to=vals.assigned_to;
  if(vals.status!==a.status) body.status=vals.status;
  if(vals.serial!==a.serial) body.serial=vals.serial;
  if(vals.vendor!==a.vendor) body.vendor=vals.vendor;
  if(vals.purchased_on!==(a.purchasedOn||'')) body.purchased_on=vals.purchased_on;
  if(vals.warranty_until!==(a.warrantyUntil||'')) body.warranty_until=vals.warranty_until;
  if(costEl && costCents!==(a.costCents!=null?a.costCents:null)){
    if(costCents==null) body.clear_cost=true; else body.cost_cents=costCents;
  }
  if(vals.note!==a.note) body.note=vals.note;
  if(Object.keys(body).length===1){ toast('Nothing changed.'); return; }
  log('Asset updated', tag+' '+name, id);
  post('PATCH','/api/assets/'+id, body)
    .then(d=>{ if(d){ if(d.version) a.version=d.version; toast(`Asset ${tag} saved.`); } });
}
